stub/src/main/java/io/grpc/stub/CallStreamObserver.java

```diff
@@ -0,0 +1,90 @@
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.stub;
+
+import io.grpc.ExperimentalApi;
+
+/**
+ * A refinement of StreamObserver provided by the GRPC runtime to the application that allows for
+ * more complex interactions with call behavior.
+ *
+ * <p>In any call there are logically two {@link StreamObserver} implementations:
+ * <ul>
+ *   <li>'inbound' - which the GRPC runtime calls when it receives messages from the
+ *   remote peer. This is implemented by the application.
+ *   </li>
+ *   <li>'outbound' - which the GRPC runtime provides to the application which it uses to
+ *   send messages to the remote peer.
+ *   </li>
+ * </ul>
+ *
+ * <p>Implementations of this interface represent the 'outbound' message se
+ *
+ */
+@ExperimentalApi
+public interface CallStreamObserver<V> extends StreamObserver<V> {
+
+  /**
+   * If {@code true} indicates that a call to {@link #onNext(Object)} will not require the entire
+   * message to be buffered before it is sent to the peer.
+   */
+  public boolean isReady();
+
+  /**
+   * Set a {@link Runnable} that will be called every time the stream {@link #isReady()} state
+   * changes from {@code false} to {@code true}.
+   *
+   * <p>Note that the handler may be called some time after {@link #isReady} has transitioned to
+   * true as other callbacks may still be executing in the 'inbound' observer.
+   *
+   * @param onReadyHandler to call when peer is ready to receive more messages.
+   */
+  public void setOnReadyHandler(Runnable onReadyHandler);
+
+  /**
+   * Disables automatic flow control where a token is returned to the peer after a call
+   * to the 'inbound' {@link io.grpc.stub.StreamObserver#onNext(Object)} has completed. If disabled
+   * an application must make explicit calls to {@link #request} to receive messages.
+   *
+   * <p>Note that for cases where the runtime knows that only one inbound message is allowed
```

Comment:
Ah, I see. Now I feel for dumb for asking since. after reading what you've written, it seems obvious in hindsight. Thanks for adding the comments.